### DISCUSSION TITLE
PDF Output seems to be only an image in a pdf?

### DISCUSSION THREAD

**Discussion by Rapid1898-code** [pymupdf/PyMuPDF#3930] [Looking for help] [answered]
Hello - i have a working solution which reads the attached teml-pdf
fillout the fields using pymupdf and then save the pdf (see also the attachd example how this finally looks like now)

It seems that the final output is only an image in the pdf (so for example i can´t select any text in the pdf).
Is there any other way to save the pdf so its not an image and a "normal" pdf which also text can be selected?

Currently i am using this statement for saving:

`out.save(f"{sheetNr}_{sheetName}.pdf", garbage=3, deflate=True) `

[11_11TH_CO_.pdf](https://github.com/user-attachments/files/17308824/11_11TH_CO_.pdf)

[templ.pdf](https://github.com/user-attachments/files/17308801/templ.pdf)

**Comment by JorjMcKie** [comment]
Your `.save()` does not / cannot **_ever_** produce a PDF with image-only pages! Something else **_must_** happen to the file thereafter.

**Comment by Rapid1898-code** [comment]
Hello - this is partly the code i am using - 
Am i doing here something wrong?
So i get this pdf which i can´t select any text?

```
import pymupdf
import xlwings as xw
import os
import sys
import time

path = os.path.abspath(os.path.dirname(sys.argv[0])) 

# open the excel-sheet and read the input-data to work on
fn = os.path.join(path, "inp.xlsx")
print(f"Try to open excel in {fn}")
wb = xw.Book (fn)
wsInp = wb.sheets[0]
inpData = wsInp.range ("C8:O47").value
inpCustomer = wsInp.range ("C2:C4").value
inpSpouse = wsInp.range ("D2:D4").value
header = [x[0] for x in inpData]

# prepare dict with all the input-data for every pdf to create
inpDict = {}
for i in range(1, 12):
  inpDict[i] = [x[i+1] for x in inpData]
# for k,v in inpDict.items():
#   print(k,v)
# exit()

skippedList = []
logInfo = []
# iterate trough all input-worksheets for which pdf should be created
for k,v in inpDict.items():
  sheetNr = int(v[1])
  sheetName = v[5].replace(" ", "_").replace(".", "_")
  print(f"Working for sheet {sheetNr}")
  # open the pdf template
  doc = pymupdf.open("templ.pdf")
  # iterate trough all pages for all fields in the template
  # fill out all the values in the pdf templates
  for page in doc:
    for field in page.widgets():
      # if field.field_value:
      #   print(field.field_name)
      #   print(field.field_value)     

      if field.field_name == "topmostSubform[0].CopyB[0].BoxA_ReadOrder[0].f2_01[0]" or \
          field.field_name == "topmostSubform[0].Copy2[0].BoxA_ReadOrder[0].f2_01[0]":
        if v[3]:
          field.field_value = v[3]
      elif field.field_name == "topmostSubform[0].CopyB[0].Col_Left[0].f2_02[0]" or \
          field.field_name == "topmostSubform[0].Copy2[0].Col_Left[0].f2_02[0]":                        
        if v[4]:
          field.field_value = v[4]
      elif field.field_name == "topmostSubform[0].CopyB[0].Col_Left[0].f2_03[0]" or \
          field.field_name == "topmostSubform[0].Copy2[0].Col_Left[0].f2_03[0]":
        if isinstance(v[10], float):
          v[10] = str(int(v[10]))
        if v[6]:
          worker = f"{v[5]}\n{v[6]}\n{v[7]}\n{v[8]}, {v[9]} {v[10]}"  
        else:
          worker = f"{v[5]}\n{v[7]}\n{v[8]}, {v[9]} {v[10]}"            
        field.field_value = worker
  
        worker = sheetOwner[:]
        worker = [x for x in worker if x]
        worker = "\n".join(worker)
        field.field_value = worker

      # write data in fields and color it
      field.text_color = [0, 0, 0]
      field.update()

  # write fille out pdf file
  out = pymupdf.open()
  for page in doc:
      w, h = page.rect.br  # page width / height taken from bottom right point coords
      outpage = out.new_page(width=w, height=h)  # out page has same dimensions
      pix = page.get_pixmap(dpi=400)  # set desired resolution
      outpage.insert_image(page.rect, pixmap=pix)
  out.save(f"{sheetNr}_{sheetName}.pdf", garbage=3, deflate=True)      
```

**Reply by JorjMcKie** [reply]
Please look at line `outpage.insert_image(page.rect, pixmap=pix)`.
You are inserting **_images_** into the new pages ...

**Comment by Rapid1898-code** [comment]
Hello - how can i change this part - so it is outputting a "normal" pdf and not images?
```

  out = pymupdf.open()
  for page in doc:
      w, h = page.rect.br  # page width / height taken from bottom right point coords
      outpage = out.new_page(width=w, height=h)  # out page has same dimensions
      pix = page.get_pixmap(dpi=400)  # set desired resolution
      outpage.insert_image(page.rect, pixmap=pix)
  out.save(f"{sheetNr}_{sheetName}.pdf", garbage=3, deflate=True)     
```

**Reply by JorjMcKie** [reply]
Instead of inserting an image, use `page.insert_pdf_page()`.

There is a caveat however:
If you you create a new PDF as you do, then it will not be a Form PDF, i.e. contain no form fields. All the fields and annotations are ignored with this method.
A PDF containing widgets / fields is a separate PDF type: a "Form PDF". This is a document-wide property and not decided on a per-page basis.

So what I recommend is simply making your changes (filling field values and what not else) and then saving it under a new name.
This should give you all you want.

**Comment by Rapid1898-code** [comment]
Hello - thx again for your help - 

No i am saving the file with a new name with this command:
`doc.save(f"{sheetNr}_{sheetName}.pdf", garbage=3, deflate=True) `

But now i have the problem that the document seems to be a "normal" pdf-file where the input-fields can be changed.
Is there no way to save this pdf - but it don´t look like a form-pdf - it should look like a "normal" pdf where nothing can be changed but the text should be able to copy paste for example.

**Reply by JorjMcKie** [reply, marked as answer]
Ok, I finally understand your intention completely I think:
1. Take a template Form PDF and give its fields values taken from ... wherever (an Excel file in your case).
2. Save the thus modified PDF in a way where the fields are visible and their text be selectable like standard text, but protected against more changes.

If correctly stated, do this:
```python
doc = pymupdf.open("templ.pdf")

# modify fields in this document as before
# ...

# now convert the Form PDF into a standard PDF
# by "baking-in" fields (and/or annotations)

doc.bake(annots=True, widgets=True)
doc.ez_save("11_11TH_CO_.pdf")
```

**Comment by Rapid1898-code** [comment]
Thx a lot - i think this is now exactly what i need and meant.
